### DISCUSSION TITLE
dGPU hot unplug?

### DISCUSSION THREAD

**Discussion by dragonnn** [YaLTeR/niri#683] [Ideas]
Hi!!
I am wondering it would be possible for niri to support nvidia dGPU hot unplug?
Hot plug works fine, I can rescan the pci bus, load nvidia modules and run 
`sudo udevadm trigger --verbose --type=devices --action=add --subsystem-match=drm --property-match="MINOR=0"` 
niri does pick up my dGPU. But when I try to run this command to make niri release the dgpu:

`sudo udevadm trigger --verbose --type=devices --action=remove --subsystem-match=drm --property-match="MINOR=0"`

my screen that is connected to the dGPU goes black, so something is happening but when I run nvidia-smi niri is still present on the dGPU process list with prevents me from rmmod the nvidia modules and then unbind the dGPU from PCI bus.

On my laptop this is necessary because the dGPU doesn't support D3 Runtime Suspend so it never goes off even when not used by any screens or programs. Currently I have to logout to be able to unbind the dGPU from PCI bus so it goes off when I am on battery. I did try the same stuff on wlroots based compositor and it didn't work too, from what I found wlroots itself should support it but they are missing bits in libglvnd to release the dGPU. Is they any hope for niri to be not limited by that?

**Comment by YaLTeR** [comment]
This might be due to niri eagerly creating egl renderers for GPUs even if no enabled outputs use that GPU. Maybe creating renderers on demand and releasing them as needed could help with this.

**Comment by dragonnn** [comment]
That sounds promising, could you point me where code that handles that is present? I don't promise anything but I am willing giving it a try to code that myself and see how it goes.

**Reply by YaLTeR** [reply]
Somewhere around here: https://github.com/YaLTeR/niri/blob/main/src%2Fbackend%2Ftty.rs#L513

I'm not sure what exactly keeps the GPU on, but I know in cosmic-comp they did something to fix this, so you may want to ask in the smithay room on matrix

**Comment by x140x1n** [comment]
any way to implement an option for niri to only use specific gpu's? I want to passthrough my dgpu with libvirt. But it's impossible to do anything with the gpu as long as niri is running. Only using the gpu for passthrough is not an option since some programs I use won't run without it.
